static/app/utils/reactRouter6Compat.tsx

```diff
@@ -227,3 +230,46 @@ export function location6ToLocation3<Q extends Query = DefaultQuery>(
     action: 'POP',
   } satisfies Location3<Q>;
 }
+
+// Shims useRouteLeave between react router versions
+export function useRouteLeave() {
+  if (USING_REACT_ROUTER_SIX) {
+    unstable_usePrompt({
+      message: 'Are you sure?',
+      when: ({currentLocation, nextLocation}) =>
+        currentLocation.pathname !== nextLocation.pathname,
+    });
+  }
+}
+
+type ReactRouterV6RouteLeaveCallback = (state: {
+  currentLocation: Location;
+  nextLocation: Location;
+}) => boolean;
+type ReactRouterV3RouteLeaveCallback = () => string | undefined;
```

Comment:
Let's add a comment that explains how the old one worked vs the new one, just so we remember what the difference is in the future without having to re-understand it